### PR TITLE
BUG: do not dereference NULL pointer

### PR DESCRIPTION
**Pull request by cgohlke**
The pointer 'current' could be NULL in case the above 'while (current != NULL) {..}' loop finishes without break.

Maybe I'm missing some logic, but better safe than sorry...

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -5178,7 +5178,7 @@ PyUFunc_RegisterLoopForDescr(PyUFuncObject *ufunc,
                 }
                 current = current->next;
             }
-            if (cmp == 0 && current->arg_dtypes == NULL) {
+            if (cmp == 0 && current != NULL && current->arg_dtypes == NULL) {
```
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: At least from the code here, it looks like one can in principle have had the last in the chain have `cmp == 0` but `last->arg_dtypes != NULL`.

Total nitpick: could you put `current != NULL` at the start?  Or perhaps this is clearer if inverted
```
if (current == NULL || !(cmp == 0 && current->arg_dtypes == NULL)) {
    result = -1;
}
else {
    ...
}
```

Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: This whole function looks to be in a poor state to me, and will return `-1` without setting an exception.

Comment (cgohlke) [reply numpy/core/src/umath/ufunc_object.c]: > will return `-1` without setting an exception

I noticed that too, but in the few cases the `PyUFunc_RegisterLoopForDescr` is used there's a if `(!PyErr_Occurred()) { PyErr_SetString(PyExc_RuntimeError...` or the output is not checked at all.

Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: added a test that reaches the "return -1", added exception string

File: numpy/core/src/umath/_struct_ufunc_tests.c.src [hunk 2]
```diff
@@ -53,6 +47,58 @@ static void add_uint64_triplet(char **args, npy_intp *dimensions,
...
+    dtypes[1] = dtype;
+    dtypes[2] = dtype;
+
+    retval = PyUFunc_RegisterLoopForDescr((PyUFuncObject *)add_triplet,
+                                dtype,
+                                &add_uint64_triplet,
+                                dtypes,
+                                NULL);
+
+    Py_DECREF(dtype);
```
Comment (seberg) [inline numpy/core/src/umath/_struct_ufunc_tests.c.src]: I think this should only happen in the error path here.

File: numpy/core/src/umath/ufunc_object.c [hunk 3]
```diff
@@ -5196,6 +5199,8 @@ PyUFunc_RegisterLoopForDescr(PyUFuncObject *ufunc,
                 current->nargs = ufunc->nargs;
             }
             else {
+                PyErr_SetString(PyExc_ValueError,
+                    "loop already registered");
```
Comment (seberg) [inline numpy/core/src/umath/ufunc_object.c]: Shouldn't this be a RuntimeError (although considering the discussion, it is good to know it didn't get replaced)?

File: numpy/core/src/umath/_struct_ufunc_tests.c.src [hunk 4]
```diff
@@ -53,6 +47,58 @@ static void add_uint64_triplet(char **args, npy_intp *dimensions,
...
+    PyArray_Descr *dtype;
+    PyArray_Descr *dtypes[3];
+    int retval;
+
+    add_triplet = PyUFunc_FromFuncAndData(NULL, NULL, NULL, 0, 2, 1,
+                                    PyUFunc_None, "add_triplet",
+                                    "add_triplet_docstring", 0);
+
+    dtype_dict = Py_BuildValue("[(s, s), (s, s), (s, s)]",
+        "f0", "u8", "f1", "u8", "f2", "u8");
```
Comment (seberg) [inline numpy/core/src/umath/_struct_ufunc_tests.c.src]: NIT:  at least 2 tabs is typical I think in numpy code base.

File: numpy/core/src/umath/_struct_ufunc_tests.c.src [hunk 5]
```diff
@@ -53,6 +47,58 @@ static void add_uint64_triplet(char **args, npy_intp *dimensions,
...
+static PyObject*
+register_fail(PyObject* NPY_UNUSED(self), PyObject* NPY_UNUSED(args))
+{
+    PyObject *add_triplet;
+    PyObject *dtype_dict;
+    PyArray_Descr *dtype;
+    PyArray_Descr *dtypes[3];
+    int retval;
+
+    add_triplet = PyUFunc_FromFuncAndData(NULL, NULL, NULL, 0, 2, 1,
```
Comment (seberg) [inline numpy/core/src/umath/_struct_ufunc_tests.c.src]: Don't you have to DECREF that before returning?

**Review by seberg — APPROVED** [review]
Well the original fix seems pretty uncontroversial? I did not check it closely. Just small comments about the test.

File: numpy/core/src/umath/_struct_ufunc_tests.c.src (hunk 2 above)
Comment (mattip) [reply numpy/core/src/umath/_struct_ufunc_tests.c.src]: right

File: numpy/core/src/umath/ufunc_object.c (hunk 3 above)
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: ok, changing

File: numpy/core/src/umath/_struct_ufunc_tests.c.src (hunk 4 above)
Comment (mattip) [reply numpy/core/src/umath/_struct_ufunc_tests.c.src]: aligning

File: numpy/core/src/umath/_struct_ufunc_tests.c.src (hunk 5 above)
Comment (mattip) [reply numpy/core/src/umath/_struct_ufunc_tests.c.src]: yes, fixing

**Comment by seberg** [comment]
From my side this can go in, unless the original fix discussion was not settled? Feel free to do the honour Matti.

**Comment by mattip** [comment]
Thanks @cgohlke

**Comment by mattip** [comment]
There was a comment about `PyUFunc_RegisterLoopForDescr` being used with no check on the return value, but I only see uses in tests, so I guess we are done here.

**Pull request merged**